### PR TITLE
bugfix: Preliminary support for 'best effort' environment installs

### LINKED ISSUE #15683: Environment/spack.yaml installs not "best effort"
(This issue is a summary of a slack exchange started with eugeneswalker on March 9, 2020.)

Running `spack install` in the e4s environment terminates after building `slate` fails despite every other package having been successfully installed and put into binary cache.  The build of `slate` is broken and a separate issue has been filed with the developer.

See https://github.com/spack/spack/pull/15415#issuecomment-597185478 for more information.

### PR DESCRIPTION
**Pull request by tldahlgren**
Fixes #15683 

The new distributed build assumes a single explicit spec while installs through `spack.yaml` files  process each spec separately in the same process.  When the installation of one of the packages in the `spack.yaml` fails, the whole process terminates.  

Catching the failure to allow the installation of subsequent packages is not sufficient for "best effort" installs since failures can trigger infinite loops in the build process of a subsequent spec.  There is an additional inefficiency in that failure markings are cleared when a new `do_install` starts so there are multiple attempts to rebuild a failing package that is a key dependency in the `spack.yaml` file.

Prefix Locking
--------------
Prefix locks cached in the database are not currently removed when the locks are released during the build process.  This can result in an infinite loop over the build queue where there is a dependency on an already installed package.

When the database pulls the prefix lock from the cache instead of creating a lock for the new `do_install`, the lock's read/write counts become inaccurate. The counts are checked when the lock for an installed package is downgraded from write-to-read resulting in a `LockDowngradeError` failure.  The installed package is then added back to the build queue to be checked again later.  This sequence continues until it is interrupted.

Solution
--------
This PR, through separate commits, addresses both the prefix lock and failure cache issues described above.  Cached prefix locks are removed when the locks are released. A `--keep-failures` install flag is added and automatically used when installing from a `spack.yaml` file.

TODO
------
- [x] Determine which of three options under consideration really addresses the locking issue
- [x] Fix existing lock-related tests

Follow-On Work
----------------
- If #15295 is merged before this PR, `environment.py` will need to be modified to support `--fast-fail`.
- If #15314 is merged before this PR, `packaging_guide.rst`'s discussion of `--keep-failures` should be changed to reference the `spack clean -f` option

### CODE REVIEW

**Comment by tldahlgren** [comment]
@eugeneswalker Does this change resolve the issue you mentioned in Spack?  If so, I can work on a better solution.

**Comment by eugeneswalker** [comment]
> @eugeneswalker Does this change resolve the issue you mentioned in Spack? If so, I can work on a better solution.

Unfortunately this did not change the result. I am in a situation where I have a big environment file. The packages needed to install the environment have all been cached, except for one. When I take this Spack environment to a fresh system, and try to install it like so:

```
$> cd directory-containing-big-spack.yaml
$> spack install --cache-only
```

It installs packages from the cache until it gets to the single package which is not in the cache. When it tries to install that one, it fails because of the `--cache-only` flag, and the whole install stops. 

If I try installing without `--cache-only`:
```
$> spack install
```

It does the same thing, except that it tries to build the package which is missing in the cache, and when that fails, the install just hangs. It doesn't exit. Just hangs. Perhaps it is waiting for a lock timeout or something ? It hangs more than 3 minutes.

@tldahlgren

**Comment by tldahlgren** [comment]
> > @eugeneswalker Does this change resolve the issue you mentioned in Spack? If so, I can work on a better solution.
> 
> Unfortunately this did not change the result. I am in a situation where I have a big environment file. The packages needed to install the environment have all been cached, except for one. When I take this Spack environment to a fresh system, and try to install it like so:
> 
> ```
> $> cd directory-containing-big-spack.yaml
> $> spack install --cache-only
> ```
> 
> It installs packages from the cache until it gets to the single package which is not in the cache. When it tries to install that one, it fails because of the `--cache-only` flag, and the whole install stops.
> 
> If I try installing without `--cache-only`:
> 
> ```
> $> spack install
> ```
> 
> It does the same thing, except that it tries to build the package which is missing in the cache, and when that fails, the install just hangs. It doesn't exit. Just hangs. Perhaps it is waiting for a lock timeout or something ? It hangs more than 3 minutes.
> 
> @tldahlgren

Ah.  I didn't realize that was the problem.

What is the best way for me to reproduce the problem?  Is there a container that readily produces the problem?

**Comment by tldahlgren** [comment]
The quick "fix" does appear to allow the installation to proceed beyond a failure; however, it does not address the problem of looping on cleanup logic.

**Comment by tldahlgren** [comment]
@eugeneswalker is going to be testing this PR on his build.

Once this is ready to be merged, I'd appreciate it if the two commits are retained.  (I've used `git rebase` multiple times to keep the changes separated as described above.

**Comment by tldahlgren** [comment]
@eugeneswalker ping

**Comment by eugeneswalker** [comment]
@tldahlgren Thanks for this. I am able to type `spack install --cache-only` in the presence of the E4S spack environment and have all the packages installed which are available in cache, and skip past those packages not in the cache.

**Comment by tldahlgren** [comment]
> @tldahlgren Thanks for this. I am able to type `spack install --cache-only` in the presence of the E4S spack environment and have all the packages installed which are available in cache, and skip past those packages not in the cache.

That's great to hear.  Thanks for checking @eugeneswalker!

File: lib/spack/spack/cmd/install.py [hunk 1]
```diff
@@ -78,6 +79,9 @@ def setup_parser(subparser):
     subparser.add_argument(
         '--overwrite', action='store_true',
         help="reinstall an existing spec, even if it has dependents")
+    subparser.add_argument(
+        '--keep-failures', action='store_true',
```
Comment (scheibelp) [inline lib/spack/spack/cmd/install.py:87]: (minor) I think `--keep-*` might be confusing because we have a couple other `--keep` arguments that control what type of cleanup this process does at the end - `--keep-failures` determines cleanup in the beginning. I think `--use-prior-failures` would be clearer.

File: lib/spack/spack/environment.py [hunk 2]
```diff
@@ -1165,21 +1165,25 @@ def install(self, user_spec, concrete_spec=None, **install_args):
...
+                fs.mkdirp(log_path)
+
+                with fs.working_dir(self.path):
+                    # Link the resulting log file into logs dir
+                    build_log_link = os.path.join(
+                        log_path, '%s-%s.log' % (spec.name, spec.dag_hash(7)))
+                    if os.path.lexists(build_log_link):
+                        os.remove(build_log_link)
+                    os.symlink(spec.package.build_log_path, build_log_link)
+        except (Exception, SystemExit) as exc:
```
Comment (scheibelp) [inline lib/spack/spack/environment.py:1222]: Two requests:
-    instead of using the try/catch here, can you use it in `install_all`? I think that since the entire function body here ends up being put into the try/catch, it would be simpler to guard the call to `_install` that occurs in `install_all`.
-   Which process ends up raising `SystemExit`? Is something in `do_install` calling `tty.die`? I think we should avoid raising that exception (and catching it). Also, is it possible to catch a more-specific exception than `Exception`?

File: lib/spack/spack/installer.py [hunk 3]
```diff
@@ -952,11 +955,16 @@ def _init_queue(self, install_deps, install_package):
...
                 otherwise ``False``
+            keep_failures (bool): ``True`` if previous failures should be kept,
+                otherwise ``False``
         """
         tty.debug('Initializing the build queue for {0}'.format(self.pkg.name))
         install_compilers = spack.config.get(
             'config:install_missing_compilers', False)
 
+        if keep_failures:
+            tty.debug('Keeping previous failure markings as requested')
```
Comment (scheibelp) [inline lib/spack/spack/installer.py]: I don't think this debug statement is necessary. Also there are occasions when the user doesn't explicitly request the failure markings (e.g. from the PR description: when installing a spec yaml file).

File: lib/spack/spack/installer.py [hunk 4]
```diff
@@ -1235,6 +1245,13 @@ def _release_lock(self, pkg_id):
                 except Exception as exc:
                     tty.warn(err.format(exc.__class__.__name__, ltype,
                                         pkg_id, str(exc)))
+            if lpkg:
+                prefix = lpkg.spec.prefix
+                if prefix:
```
Comment (scheibelp) [inline lib/spack/spack/installer.py]: When does `bool(prefix) == False`? I thought it always exists. Is this intended to check whether the prefix actually exists on the filesystem?

File: lib/spack/spack/test/cmd/env.py [hunk 5]
```diff
@@ -149,6 +150,88 @@ def test_env_install_all(install_mockery, mock_fetch):
...
+    previous spec fails.  In this case the dependency and dependent are
+    provided, in order, so the dependency will fail once when built explicitly
+    and, again, when it is a dependency.
+    """
+    err_msg = 'Mock _install_task error'
+
+    # Save original _install_task function for conditional use in monkeypatch
+    orig_fn = spack.installer.PackageInstaller._install_task
+
+    def _inst_task(inst, task, **kwargs):
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/env.py:167]: (minor) I have a complaint about this similar to https://github.com/spack/spack/pull/15295#discussion_r388030818

That being said I consider it less of an issue because `PackageInstaller` itself isn't being tested here.

File: lib/spack/spack/test/cmd/env.py [hunk 6]
```diff
@@ -149,6 +150,88 @@ def test_env_install_all(install_mockery, mock_fetch):
...
+    """Test install_all when a successfully installed package is repeatedly
+    installed.
+
+    This test uses the environment installation process to exercise the
+    distributed build multiple times with overlapping dependencies in the
+    same process to ensure proper management of package installation statuses.
+
+    A key part of this test addresses a problem that occurred when attempting
+    to install a package dependent on another successfully installed, multiply
+    referenced package.  An infinite loop would arise after an earlier package
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/env.py]: This combination of features is difficult for me to visualize: could you add an ascii diagram here? Something like:

```
  W
 / \
X.  Y
 \ /
  Z
```

With an explanation like "Z succeeds, but then Y fails..."

Also, you are using the diamond DAG mock package. To make sure, is that the simplest DAG this problem can occur for? For example could it occur for a DAG like

```
W
|
X
|
Z
```

?

File: lib/spack/spack/installer.py [hunk 7]
```diff
@@ -1235,6 +1245,13 @@ def _release_lock(self, pkg_id):
                 except Exception as exc:
                     tty.warn(err.format(exc.__class__.__name__, ltype,
                                         pkg_id, str(exc)))
+            if lpkg:
```
Comment (scheibelp) [inline lib/spack/spack/installer.py]: It looks like `bool(lpkg) == True` whenever `bool(lock) == True`. If I'm correct, you can remove `if lpkg:`.

File: lib/spack/spack/installer.py [hunk 8]
```diff
@@ -881,7 +883,7 @@ def _ensure_locked(self, lock_type, pkg):
             .format(lock_type)
 
         pkg_id = package_id(pkg)
-        ltype, lock = self.locks.get(pkg_id, (lock_type, None))
+        ltype, lpkg, lock = self.locks.get(pkg_id, (lock_type, None, None))
```
Comment (scheibelp) [inline lib/spack/spack/installer.py]: `lpkg` value is not used in this function. When you extract a tuple into multiple values, you can discard one with `_` like:

```
ltype, _, lock = self.locks.get(...)
```

File: lib/spack/spack/installer.py [hunk 9]
```diff
@@ -932,6 +934,7 @@ def _ensure_locked(self, lock_type, pkg):
         except (lk.LockDowngradeError, lk.LockTimeoutError) as exc:
             tty.debug(err.format(op, desc, pkg_id, exc.__class__.__name__,
```
Comment (scheibelp) [inline lib/spack/spack/installer.py:966]: (main question) In the case of a `LockDowngradeError` I think that implies a bug in Spack - in that case it should probably terminate. Is it clear why this occurs and/or what the read/write counts are at the time of failure?

File: lib/spack/spack/test/cmd/env.py (hunk 6 above)
Comment (scheibelp) [reply lib/spack/spack/test/cmd/env.py]: Based on our discussion I think the issue looks like installing DAGs that overlap, for example

```
X
|
Z
```

followed by 

```
Y
|
Z
```

where both of these DAGs are listed as separate entries in an environment spack.yaml file, and you install all the individual specs in the environment.

File: lib/spack/spack/cmd/install.py (hunk 1 above)
Comment (tgamblin) [reply lib/spack/spack/cmd/install.py:87]: I find the wording for this one confusing... I think `--keep-failures` is easier to remember and more consistent (despite the difference in time).  Can we just make sure the `help` strings are clear?

**Comment by tgamblin** [comment]
I looked through this one and didn't have any questions Peter hadn't already asked.  I made one comment.

File: lib/spack/spack/installer.py (hunk 8 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py]: I couldn't recall if this project uses that convention.  I'll try to keep that in mind for the future too.

File: lib/spack/spack/cmd/install.py (hunk 1 above)
Comment (tldahlgren) [reply lib/spack/spack/cmd/install.py:87]: The help string currently reads `don't remove previous install failure marks before installation`.

Comment (scheibelp) [reply lib/spack/spack/cmd/install.py:87]: That's OK with me, it's not critical to change it.

File: lib/spack/spack/installer.py (hunk 8 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py]: Replaced.

File: lib/spack/spack/installer.py (hunk 3 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py]: Fair enough.  Removed.

File: lib/spack/spack/installer.py (hunk 4 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py]: Removed check

File: lib/spack/spack/test/cmd/env.py (hunk 5 above)
Comment (tldahlgren) [reply lib/spack/spack/test/cmd/env.py:167]: Noted.  I appreciate it is less of an issue since the package being tested is not `PackageInstaller`.

File: lib/spack/spack/installer.py (hunk 7 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py]: Removed.

File: lib/spack/spack/installer.py (hunk 9 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py:966]: > (main question) In the case of a `LockDowngradeError` I think that implies a bug in Spack - in that case it should probably terminate. Is it clear why this occurs and/or what the read/write counts are at the time of failure?

It was my impression that the read/write counts are intended primarily for tracking nested locking for a process.  In the installation context for a spec, they are generally 0 or 1.  In the context of an environment installation with a broken package that assumption appears to break down.  

I don't have notes on exactly what the counts were but I do recall that they increased with each iteration on a package that had a cached lock in place.  (For some reason I'm thinking the package was `hypre`.)

The problem with terminating on the `LockDowngradeError` failure is we don't then perform a 'best effort' installation in the environment context, which is the whole point of this PR.

Comment (scheibelp) [reply lib/spack/spack/installer.py:966]: > In particular, IIRC the (broken) slate package within the aforementioned environment was causing a SystemExit that precluded lock cleanup. So when a package

I haven't parsed this fully yet but this line appears incomplete

Comment (tldahlgren) [reply lib/spack/spack/installer.py:966]: > > In particular, IIRC the (broken) slate package within the aforementioned environment was causing a SystemExit that precluded lock cleanup. So when a package
> 
> I haven't parsed this fully yet but this line appears incomplete

Thanks for catching that.  That might have been a result of some of the "help" one of the foster kittens was providing when he walked on the laptop keyboard. :)

File: lib/spack/spack/test/cmd/env.py (hunk 6 above)
Comment (tldahlgren) [reply lib/spack/spack/test/cmd/env.py]: > Based on our discussion I think the issue looks like installing DAGs that overlap, for example
> 
> ```
> X
> |
> Z
> ```
> 
> followed by
> 
> ```
> Y
> |
> Z
> ```
> 
> where both of these DAGs are listed as separate entries in an environment spack.yaml file, and you install all the individual specs in the environment.

You are correct that the diamond pattern is not the simplest case where this can occur.  I was trying to re-use existing packages.  I added the example below to the test code.  Given packages `A`, `B`, and `Depb`, where `A` and `Depb` depend on `B`:

```
    A      Depb
    |       |
    B       B
```

If the installation of `B` succeeds while `A` fails, then the installation of `Depb` should still be allowed to succeed.

**Comment by tldahlgren** [comment]
@scheibelp ping

File: lib/spack/spack/environment.py (hunk 2 above)
Comment (scheibelp) [reply lib/spack/spack/environment.py:1222]: @tldahlgren this comment appears to have been edited strangely: is this the end result you intended given the starting state of the comment?

File: lib/spack/spack/test/cmd/env.py [hunk 10]
```diff
@@ -149,6 +150,95 @@ def test_env_install_all(install_mockery, mock_fetch):
...
+    out = capfd.readouterr()[1]
+    env_specs = e._get_environment_specs()
+    for spec in env_specs:
+        assert not spec.package.installed
+    assert out.count(err_msg) == 1  # b_explicit, not as b_dependency
+    assert 'Skipping build of a' in out
+
+
+def test_env_install_all_seq(install_mockery, mock_fetch, monkeypatch):
+    """Test install_all when a successfully installed package is a dependent
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/env.py:192]: I think you mean "dependency" here where you say "dependent" (there is no single package in this example which depends on multiple specs, but there is a package which is depended-on by multiple specs.

File: lib/spack/spack/test/cmd/env.py [hunk 11]
```diff
@@ -149,6 +150,95 @@ def test_env_install_all(install_mockery, mock_fetch):
...
+    of multiple specs the earlier of which fails to install.
+
+    This test uses the environment installation process to exercise the
+    distributed build multiple times with overlapping dependencies in the
+    same process to ensure proper management of package installation statuses.
```
Comment (scheibelp) [inline lib/spack/spack/test/cmd/env.py:197]: > ensure proper management of package installation statuses

> This test requires the environment perform a "best effort" installation

I think I would rephrase this like

> Ensure that environments installing a collection of specs perform a "best effort" installation. For example, given two packages added in order, A and Depb...

File: lib/spack/docs/packaging_guide.rst [hunk 12]
```diff
@@ -4315,6 +4315,24 @@ to get rid of the install prefix before you build again:
 
    $ spack uninstall --force <spec>
 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Keeping install failure tracking data
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+By default, ``spack install`` will clear all persistent install
+failure tracking information as part of its set up process. If you
```
Comment (scheibelp) [inline lib/spack/docs/packaging_guide.rst:4323]: (minor) Would it make sense to mention that normally Spack will try to reinstall each (not-yet-installed) package every time you invoke `spack install`?

File: lib/spack/spack/environment.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/environment.py:1222]: My response at the time was only to that part of the second bullet.  I keep forgetting to find and respond to the original bullet points.

Comment (tldahlgren) [reply lib/spack/spack/environment.py:1222]: Uh oh.  I now see the issue.  I accidentally edited your original comment instead of quote replying.  (Sorry.)   It should be restored now.

Comment (tldahlgren) [reply lib/spack/spack/environment.py:1222]: > Two requests:
> 
>     * instead of using the try/catch here, can you use it in install_all? I think that since the entire function body here ends up being put into the try/catch, it would be simpler to guard the call to _install that occurs in install_all.

I suspect you are correct.  Let me look at that again.


> 
>     * Which process ends up raising SystemExit? Is something in do_install calling tty.die? I think we should avoid raising that exception (and catching it). 

IIRC It was the installation of the `slate` package -- which needed to be fixed -- that was raising `SystemExit` and precluded lock cleanup.


>Also, is it possible to catch a more-specific exception than Exception?

I'll have to investigate this further.

File: lib/spack/spack/installer.py (hunk 9 above)
Comment (scheibelp) [reply lib/spack/spack/installer.py:966]: If the locks are reentrant and allow for counts that are above 1. Then perhaps the logic in `llnl.util.lock` should be updated. For example

```
    def downgrade_write_to_read(self, timeout=None):
 
        if self._writes == 1 and self._reads == 0:
```

could change to 

```
    def downgrade_write_to_read(self, timeout=None):
 
        if self._writes == 1:  # no need to count how many read locks there are
```

?

File: lib/spack/spack/environment.py (hunk 2 above)
Comment (tldahlgren) [reply lib/spack/spack/environment.py:1222]: > > Two requests:
> > ```
> > * instead of using the try/catch here, can you use it in install_all? I think that since the entire function body here ends up being put into the try/catch, it would be simpler to guard the call to _install that occurs in install_all.
> > ```
> 
> I suspect you are correct. Let me look at that again.

I added the `try`/`catch` block in `_install` because `_install` is called from `install` and `install_all`.  @scheibelp Are you suggesting I move the block to both calls?

Comment (tldahlgren) [reply lib/spack/spack/environment.py:1222]: > 
> > Also, is it possible to catch a more-specific exception than Exception?
> 
> I'll have to investigate this further.

I used `Exception` here since it has been used by the package install process.

File: lib/spack/spack/installer.py (hunk 9 above)
Comment (tldahlgren) [reply lib/spack/spack/installer.py:966]: > If the locks are reentrant and allow for counts that are above 1. Then perhaps the logic in `llnl.util.lock` should be updated. For example
> 
> ```
>     def downgrade_write_to_read(self, timeout=None):
>  
>         if self._writes == 1 and self._reads == 0:
> ```
> 
> could change to
> 
> ```
>     def downgrade_write_to_read(self, timeout=None):
>  
>         if self._writes == 1:  # no need to count how many read locks there are
> ```
> 
> ?

Perhaps.  I was precluding nested transactions for the downgrade.  I'd appreciate @tgamblin 's perspective on this since he's the locking expert.

File: lib/spack/spack/environment.py (hunk 2 above)
Comment (scheibelp) [reply lib/spack/spack/environment.py:1222]: > I added the try/catch block in _install because _install is called from install and install_all. @scheibelp Are you suggesting I move the block to both calls?

It would only need to be caught in `install_all` though, wouldn't it? The problem was the installation of multiple root specs. If it in fact would need to be caught in both functions, then I agree that it's best to catch the exception here.

> I used Exception here since it has been used by the package install process.

As in there is code there which raises a generic `Exception`? I'd assume most parts of Spack raise `SpackError`.

Also, if there are packages which raise a `SystemExit`, then I'm inclined to say we should allow it to terminate the program. If people want different behavior, they should throw a different exception.

**Comment by mrmundt** [comment]
Curious party - is this likely to be in the next release?

**Comment by tgamblin** [comment]
> Curious party - is this likely to be in the next release?

yes!

**Comment by mrmundt** [comment]
> > Curious party - is this likely to be in the next release?
> 
> yes!

Great! Thank you, @tgamblin ! Is there an ETA for that release?

**Comment by tldahlgren** [comment]
Replaced by #18131
